Adding infrastructure for retrying requests when they fail

Alexalex1432 opened the pull request:
Only on the final attempt adding error logs for user debugging.

This can help address situations where the failures are caused by race conditions or time-based issues where a record is locked.  Scheduled jobs will pick up the record of the sync and retry before generating errors to the user

Commit:
Adding infrastructure for retrying requests when they fail, and only on the final attempt adding error logs for user debugging.

This can help address situations where the failures are caused by race conditions or time-based issues where a record is locked.  Scheduled jobs will pick up the record of the sync and retry before generating errors to the user
Resolving FUN-11260

adelorme82 commented on src/classes/frCampaignTest.cls:
```diff
@@ -48,17 +48,19 @@ public class frCampaignTest {
         request.put('goalId', goalId);
-        request.put('deleted', false);
+        
+        RestRequest req = new RestRequest();
```
I see this block a lot, thoughts on moving it into a helper method? Can accept a route and request, and set all the global things accordingly

adelorme82 commented on src/classes/frCampaignTest.cls:
```diff
@@ -71,31 +73,31 @@ public class frCampaignTest {
-        childRequest.put('deleted', false);
-        
-        new frCampaign(parentRequest);
+        Map<String, Object> request = new Map<String, Object>();
```
Can this Map start as `getTestRequest()` so it's more clear what things you are setting that are different than normal?

adelorme82 commented on src/classes/frCampaignTest.cls:
```diff
@@ -169,17 +195,25 @@ public class frCampaignTest {
-        parentRequest.put('parentGoalId', topLevelGoalId);
-        parentRequest.put('goalId', parentGoalId);
-        parentRequest.put('deleted', true);
+        Map<String, Object> request = new Map<String, Object>();
```
Same here, can it start as `getTestRequest()`

adelorme82 commented on src/classes/frCampaignTest.cls:
```diff
@@ -274,46 +332,43 @@ public class frCampaignTest {
-        new frCampaign(grandchildCampaign1);
-        new frCampaign(grandchildCampaign2);
+        Campaign grandchildCampaign1 = new Campaign(
+        	Name = 'Child campaign 1',
```
Super nitpicky but there are some tabs here that look foreign amongst all the correct whitespace characters

adelorme82 commented on src/classes/frDataMigration.cls:
```diff
@@ -37,8 +37,8 @@
 */
-public class frDataMigration implements InstallHandler {
-    public void onInstall(InstallContext context) {
+public class frDataMigration {
```
Do we still care about this class?

Alexalex1432 replied:
It's basically our holder class for any evolutions.  It was once populated with some migrations, but we deployed that and had to remove the migration code after it ran because it referenced a field that we wanted to delete (which was the reason for the migration).  It would take us some effort to remove the class from the managed package, so I figure we can keep it around in case we have to do any data migrations going forward.  

Alternatively, we could also just have frPostInstall handle that now since that is what calls this

adelorme82 commented on src/classes/frDonationTest.cls:
```diff
@@ -64,19 +64,9 @@ public class frDonationTest {
-        
-        MockResponse response = (MockResponse) JSON.deserialize(res.responseBody.toString(), MockResponse.class);
-        
-        List<Error__c> errors = [SELECT Id, Name, Error__c from Error__c];
```
Why do we no longer have an error anywhere?

Alexalex1432 replied:
Errors are now only generated on the final attempt of a sync (right now that's after 3 attempts).  So the logic these tests are executing are the first sync attempt, the only way to get Error__c logs generated is through the scheduled job picking up an attempt on its last try

adelorme82 commented on src/classes/frDonorEmails.cls:
```diff
@@ -1,41 +1,44 @@
+                    emr.RelationId = contacts.get(0).Id;
+                    emr.RelationType = 'ToAddress';
+                    insert emr;
+					result = true;
```
:upside_down_face:

adelorme82 commented on src/classes/frSyncRequestHandler.cls:
```diff
@@ -0,0 +1,74 @@
+        List<Sync_Attempt__c> toUpdate = new List<Sync_Attempt__c>();
+        for (Sync_Attempt__c attempt : syncs) {
+            attempt.Attempts__c += 1;
+            Boolean success = sync(attempt);
```
Can `sync` throw an exception? Should we try/catch this to make sure that the `Attempts__c` field is *always* incremented?

Commit:
Refactoring creation of RestRequest in tests to use a common utility method to reduce a lot of repeated code

Additionally fixing some whitespace
It also appears the requestURI is not important in tests, so dropping it.
Resolving FUN-11260

Commit:
Adding handling for exceptions in frSyncRequestHandler.  Ensure that individually the records attempts are incremented in an exception, the exception gets propagated (leading to an email to package maintainers)

Resolving FUN-11260
